File: src/rubik/tests/shapes.py
#!/usr/bin/env python3
import logging
import unittest

import numpy as np
from numpy import testing

from rubik import shapes
from rubik.colours.default_colours import Colours
from common.variables import variable_names_and_objects
import itertools
from common.cli import log, unit_test_parse
from common.arguments import all_combinations


class BasicProperties(unittest.TestCase):

    def setUp(self):
        self.kwargs = {"N": 3, "M": 4}
        self.args = []
        self.names_and_shapes_classes = variable_names_and_objects(shapes.Tile,
                                                                   shapes.Domino,
                                                                   shapes.Strip,
                                                                   shapes.Square,
                                                                   shapes.Grid,
                                                                   shapes.Sheet,
                                                                   shapes.Volume,
                                                                   shapes.Cube,
                                                                   shapes.Triangle,
                                                                   shapes.Tetrahedron,
                                                                   )

    def test_printing(self):
        """Should print a nicely coloured shape."""
        for name, shape_class in self.names_and_shapes_classes:
            log.debug(f"\nTrying to print a {name}")
            shape = shape_class(*self.args, **self.kwargs)
            log.debug(f"{shape!r}")
            log.debug(f"{shape}")
            log.debug(f"{shape = }")
            log.debug(f"The solved configuration is: {shape.solved_config(*self.args, **self.kwargs)}, and should look like {shape.clean_config(*self.args, **self.kwargs)}")
            self.assertEqual(shape, shape.clean_config(*self.args, **self.kwargs), "Shapes must always start in a clean configuration.")

    def test_unique_row_containers(self):
        """
        Some operations will be swapping rows and columns,
        and when we swap rows, we need to ensure the
        containers are different, and we don't have
        two duplicate rows where swapping elements in
        one would inadvertently affect the contents of the
        other. Hence we need each container to be unique.
        """

        def traverse_rows(faces, function):
            """Traverse the rows."""
            for face in faces:
                for row in face:
                    yield function(row)

        for name, shape_class in self.names_and_shapes_classes:
            shape = shape_class(*self.args, **self.kwargs)
            ids = list(traverse_rows(shape._faces, id))
            self.assertEqual(len(ids), len(set(ids)), f"There are duplicate memory locations used to store the tiles in {name}. Not suitable for swapping operations.")


class Moves(unittest.TestCase):

    def test_single_moves_domino(self):
        shape = shapes.Domino()
        for move in shape.moves():
            shape_moved = shape.move(move)
            self.assertEqual(shape, shape.solved_config())
            self.assertNotEqual(shape_moved, shape)
            self.assertNotEqual(shape_moved, shape.solved_config())
            shape_reverted = shape_moved.move(move, reverse=True)
            self.assertEqual(shape_reverted, shape.solved_config())
            self.assertEqual(shape_reverted, shape)

    def test_single_moves_sheet(self):
        shape = shapes.Sheet()
        for move in shape.moves():
            shape_moved = shape.move(move)
            self.assertEqual(shape, shape.solved_config())
            self.assertNotEqual(shape_moved, shape)
            self.assertNotEqual(shape_moved, shape.solved_config())
            shape_reverted = shape_moved.move(move, reverse=True)
            self.assertEqual(shape_reverted, shape.solved_config())
            self.assertEqual(shape_reverted, shape)

    def test_double_moves_domino(self):
        shape = shapes.Domino()
        for move_1, move_2 in itertools.product(*[shape.moves() for i in range(2)]):
            shape_moved_once = shape.move(move_1)
            shape_moved_twice = shape_moved_once.move(move_2)
            log.debug(f"original = {shape} moved_once = {shape_moved_once} moved_twice = {shape_moved_twice}")
            self.assertEqual(shape_moved_twice, shape.solved_config())
            self.assertNotEqual(shape_moved_twice, shape_moved_once)
            shape_moved_twice_reverted = shape_moved_twice.move(move_2, reverse=True)
            self.assertEqual(shape_moved_twice_reverted, shape_moved_once)
            shape_reverted_in_order = shape_moved_twice.move(move_2, move_1, reverse=True)
            self.assertEqual(shape_reverted_in_order, shape)
            shape_reverted_out_of_order = shape_moved_twice.move(move_1, move_2, reverse=True)
            self.assertEqual(shape_reverted_out_of_order, shape)

    def test_same_moves_compare_equal(self):
        shape = shapes.Volume()
        for id_1, move_1 in enumerate(shape.moves()):
            for id_2, move_2 in enumerate(shape.moves()):
                if id_1 == id_2:
                    self.assertEqual(move_1, move_2)
                else:
                    self.assertNotEqual(move_1, move_2)

    def test_double_moves_volume(self):
        shape = shapes.Volume()
        for direction in [True, False]:
            other_direction = not direction
            for move_1, move_2 in itertools.product(*[shape.moves() for i in range(2)]):
                shape_moved_once = shape.move(move_1, reverse=direction)
                shape_moved_twice = shape_moved_once.move(move_2, reverse=direction)
                self.assertNotEqual(shape_moved_twice, shape.solved_config())
                self.assertNotEqual(shape_moved_twice, shape_moved_once)
                shape_moved_twice_reverted = shape_moved_twice.move(move_2, reverse=other_direction)
                self.assertEqual(shape_moved_twice_reverted, shape_moved_once)
                shape_reverted_in_order = shape_moved_twice.move(move_2, move_1, reverse=other_direction)
                self.assertEqual(shape_reverted_in_order, shape)
                shape_reverted_out_of_order = shape_moved_twice.move(move_1, move_2, reverse=other_direction)
                if move_1 == move_2:
                    self.assertEqual(shape_reverted_out_of_order, shape)
                else:
                    self.assertNotEqual(shape_reverted_out_of_order, shape)

    def test_rotational_symmetry(self):
        shape = shapes.Volume()
        for direction in [True, False]:
            other_direction = not direction
            for move in shape.moves():
                # The two-fold symmetry
                moved = shape.move(move, move, reverse=direction)
                moved_reverse = shape.move(move, move, reverse=other_direction)
                self.assertEqual(moved, moved_reverse)
                # The four-fold symmetry
                moved_fully = shape.move(move, move, move, move, reverse=direction)
                self.assertEqual(shape, moved_fully)

    def test_shuffle(self):
        for turns in range(10):
            shape = shapes.Volume()
            shuffled, path = shape.shuffle(turns=turns)
            log.info(f"The target shuffled cube is: {shuffled}")
            log.debug(f"Obtained by:")
            for turn, (move, reverse) in enumerate(zip(path.moves, path.reverses)):
                log.debug(f"{turn = }: {move} {'in reverse' if reverse else ''}")
            moved = shape
            log.info(f"The starting configuration is: {moved}")
            for turn, (move, reverse) in enumerate(zip(path.moves, path.reverses)):
                log.debug(f"{turn = }: {move} {'in reverse' if reverse else ''}")
                moved = moved.move(move, reverse=reverse)
                log.debug(f"The moved configuration is: {moved}")
            log.info(f"The final moved configuration is: {moved}")
            log.info(f"The target configuration is: {shuffled}")
            self.assertEqual(moved, shuffled, f"We should have recovered our target configuration.")
            reverted = moved
            for turn, (move, reverse) in reversed(list(enumerate(zip(path.moves, path.reverses)))):
                other_direction = not reverse
                log.debug(f"{turn = }: {move} {'in reverse' if other_direction else ''}")
                reverted = reverted.move(move, reverse=other_direction)
            self.assertEqual(reverted, shape)

    def test_moves_give_new_objects(self):
        shape = shapes.Volume()
        for moves in all_combinations(shape.moves()):
            for reverse in [True, False]:
                self.assertIsNot(shape, shape.move(*moves, reverse=reverse))

    def test_shuffles_give_new_objects(self):
        shape = shapes.Volume()
        for turns in range(5):
            self.assertIsNot(shape, shape.shuffle(turns))

    def test_shuffle_seeding(self):
        shape_1 = shapes.Volume()
        shape_2 = shapes.Volume()
        for seed in [False, 0]:
            while True:
                shuffle_1, path_1 = shape_1.shuffle(seed=seed)
                shuffle_2, path_2 = shape_2.shuffle(seed=seed)
                if shuffle_1 != shape_1 and shuffle_2 != shape_2:
                    break
            if seed or (isinstance(seed, int) and type(seed) != bool):
                self.assertEqual(path_1, path_2)
                self.assertEqual(shuffle_1, shuffle_2)
            else:
                self.assertNotEqual(path_1, path_2)


class Casting(unittest.TestCase):

    def test_array_casting(self):
        shape = shapes.Volume()
        faces_array = shape.to_array()
        self.assertIsInstance(faces_array, np.ndarray)
        self.assertEqual(len(faces_array.shape), 1, f"Our array must be one dimensional.")
        self.assertGreaterEqual(faces_array.size, 1, f"Our array must contain some elements.")
        shape_recovered = shape.from_array(array=faces_array)
        self.assertIsNot(shape_recovered, shape, f"These should be separate objects.")
        self.assertEqual(shape_recovered, shape, f"The shape recovered from the array conversion should match the original.")
        testing.assert_array_equal(faces_array, shape.to_array())
        faces_array[0] += 1
        with testing.assert_raises(AssertionError):
            testing.assert_array_equal(faces_array, shape.to_array(), f"Modifying the return should not change the original.")


if __name__ == '__main__':
    unit_test_parse()
    unittest.main()

File: src/rubik/shapes/tests/shapes.cpp
#include "testing/testing.h"
#include "logging/logging.hpp"
#include "rubik/shapes/volume.hpp"
#include "rubik/shapes/cube.hpp"
#include "rubik/shapes/shape.hpp"
#include <ranges>

template<typename S>
class ShapeTest : public testing::Test {
};

template<typename S>
class ShapeTestPrivate : public ShapeTest<S>, public S /*cf: https://stackoverflow.com/a/24445370/5134817*/ {
};


using ShapeTypes = ::testing::Types<Volume, Cube>;
TYPED_TEST_SUITE(ShapeTest, ShapeTypes);
TYPED_TEST_SUITE(ShapeTestPrivate, ShapeTypes);

TYPED_TEST(ShapeTest, printing) {
    TypeParam shape;
    LOG_DEBUG << shape;
}

TYPED_TEST(ShapeTest, equality) {
    TypeParam shape_1, shape_2;
    ASSERT_EQ(shape_1, shape_2);
}

TYPED_TEST(ShapeTest, hashing) {
    TypeParam shape_1, shape_2;
    ASSERT_EQ(std::hash<TypeParam>{}(shape_1), std::hash<TypeParam>{}(shape_2));
}

TYPED_TEST(ShapeTest, after_move) {
    TypeParam shape_1, shape_2;
    for (auto move: shape_1.moves()) {
        shape_2 = shape_1.move(move);
        ASSERT_NE(shape_1, shape_2);
        ASSERT_NE(std::hash<TypeParam>{}(shape_1), std::hash<TypeParam>{}(shape_2));
    }
}

TYPED_TEST(ShapeTest, moves_equality) {
    TypeParam shape;
    auto moves = shape.moves();
    for (size_t i = 0; i < moves.size(); i++) {
        for (size_t j = 0; j < moves.size(); j++) {
            if (i == j) {
                ASSERT_EQ(moves.at(i), moves.at(j));
            } else {
                ASSERT_NE(moves.at(i), moves.at(j));
            }
        }
    }
}

TYPED_TEST(ShapeTest, single_moves) {
    TypeParam shape_1, shape_2, shape_3;
    for (auto move: shape_1.moves()) {
        shape_3 = shape_2.move(move);
        ASSERT_NE(shape_3, shape_2);
        ASSERT_EQ(shape_1, shape_2);
    }
}

TYPED_TEST(ShapeTest, reverse_moves) {
    TypeParam shape_1, shape_2, shape_3;
    for (auto move: shape_1.moves()) {
        shape_2 = shape_1.move(move);
        bool reverse = true;
        shape_3 = shape_2.move(move, reverse);
        ASSERT_NE(shape_1, shape_2);
        ASSERT_EQ(shape_1, shape_3);
    }
}

TYPED_TEST(ShapeTest, compound_moves) {
    TypeParam shape_1, shape_2, shape_3, shape_4;
    for (auto reverse: {true, false}) {
        for (auto move_1: shape_1.moves()) {
            for (auto move_2: shape_1.moves()) {
                shape_2 = shape_1.move(move_1, reverse);
                shape_3 = shape_2.move(move_2, reverse);
                shape_4 = shape_1.move({move_1, move_2}, reverse);
                ASSERT_EQ(shape_3, shape_4);
            }
        }
    }
}

TYPED_TEST(ShapeTest, multiple_moves) {
    TypeParam shape_1, shape_2, shape_3;
    for (auto reverse: {true, false}) {
        for (auto move_1: shape_1.moves()) {
            for (auto move_2: shape_1.moves()) {
                shape_2 = shape_1.move({move_1, move_2}, reverse);
                shape_3 = shape_1.move(move_1, reverse).move(move_2, reverse);
                ASSERT_EQ(shape_2, shape_3);
            }
        }
    }
}

TYPED_TEST(ShapeTest, double_moves) {
    TypeParam shape;
    for (auto direction: {true, false}) {
        auto other_direction = not direction;
        for (auto move_1: shape.moves()) {
            for (auto move_2: shape.moves()) {
                auto shape_moved_once = shape.move(move_1, direction);
                auto shape_moved_twice = shape_moved_once.move(move_2, direction);
                if (shape.reverse_of(move_1) == move_2) {
                    ASSERT_EQ(shape_moved_twice, shape.solved_config());
                } else {
                    ASSERT_NE(shape_moved_twice, shape.solved_config());
                    ASSERT_NE(shape_moved_twice, shape_moved_once);
                }
                auto shape_moved_twice_reverted = shape_moved_twice.move(move_2, other_direction);
                ASSERT_EQ(shape_moved_twice_reverted, shape_moved_once);
                auto shape_reverted_in_order = shape_moved_twice.move({move_2, move_1}, other_direction);
                ASSERT_EQ(shape_reverted_in_order, shape);
                auto shape_reverted_out_of_order = shape_moved_twice.move({move_1, move_2}, other_direction);
                if (move_1 == move_2 or shape.reverse_of(move_1) == move_2 or shape.commutative(move_1, move_2)) {
                    ASSERT_EQ(shape_reverted_out_of_order, shape);
                } else {
                    ASSERT_NE(shape_reverted_out_of_order, shape)
                                                << "Performing the following moves in reverse in the wrong order should not recover the original."
                                                << "\nmove_1: " << move_1
                                                << "\nmove_2: " << move_2;
                }

            }
        }
    }
}

TYPED_TEST(ShapeTest, rotational_symmetry) {
    TypeParam shape;
    for (auto direction: {true, false}) {
        auto other_direction = not direction;
        for (auto move: shape.moves()) {
            // The two-fold symmetry
            auto moved = shape.move({move, move}, direction);
            auto moved_reverse = shape.move({move, move}, other_direction);
            ASSERT_EQ(moved, moved_reverse);
            // The four-fold symmetry
            auto moved_fully = shape.move({move, move, move, move}, direction);
            ASSERT_EQ(shape, moved_fully);
        }
    }
}

TYPED_TEST(ShapeTest, shuffle) {
    for (size_t turns = 0; turns < 10; turns++) {
        TypeParam shape;
        auto [shuffled, path] = shape.shuffle(turns, std::nullopt);
        LOG_INFO << "The target shuffled cube is:" << shuffled;
        LOG_DEBUG << "Obtained by:" << path;
        auto moved = shape;
        LOG_DEBUG << "The starting configuration is: " << moved;
        size_t turn = 0;
        for (auto [move, reverse]: path) {
            LOG_DEBUG << "Turn = " << turn++ << move << (reverse ? " in reverse" : "");
            moved = moved.move(move, reverse);
            LOG_DEBUG << "The moved configuration is: " << moved;
        }
        LOG_INFO << "The final moved configuration is: " << moved;
        LOG_INFO << "The target configuration is: " << shuffled;
        ASSERT_EQ(moved, shuffled) << "We should have recovered our target configuration.";
        auto reverted = moved;
        for (auto [move, other_direction]: path.reversed()) {
            LOG_DEBUG << "Turn = " << turn-- << move << (other_direction ? " in reverse" : "");
            reverted = reverted.move(move, other_direction);
        }
        ASSERT_EQ(reverted, shape);
    }
}

TYPED_TEST(ShapeTest, shuffle_seeding) {
    TypeParam shape_1, shape_2, shuffle_1, shuffle_2;
    Path path_1, path_2;
    auto turns = 10;
    const std::vector<typename TypeParam::Seed> seeds = {std::nullopt, 0, 1, 2};
    for (const auto &seed: seeds) {
        while (true) {
            std::tie(shuffle_1, path_1) = shape_1.shuffle(turns, seed);
            std::tie(shuffle_2, path_2) = shape_2.shuffle(turns, seed);
            if (shuffle_1 != shape_1 and shuffle_2 != shape_2) {
                break;
            }
        }

        if (seed.has_value()) {
            ASSERT_EQ(path_1, path_2);
            ASSERT_EQ(shuffle_1, shuffle_2);
        } else {
            ASSERT_NE(path_1, path_2);
        }
    }
}

TYPED_TEST(ShapeTestPrivate, invariant_positions) {

    auto check_invariant_positions = [](const auto &shape) {
        auto solved_config = shape.solved_config();
        for (const auto &[face, row, column]: shape.invariant_tile_positions()) {
            std::array<Tile, 2> tiles;
            for (size_t t = 0; const auto &s: {shape, solved_config}) {
                tiles.at(t++) = TypeParam::faces_of(s).at(face).at(row).at(column);
            }
            CHECK_EQ(tiles.front(), tiles.back()) << "The tiles dont match for face = " << face << " row = " << row
                                                  << " column = " << column;
        }
    };

    TypeParam shape;
    auto seed = 0;
    for (auto turns = 0; turns < 10; turns++) {
        auto [shuffled, path] = shape.shuffle(turns, seed);
        check_invariant_positions(shuffled);
    }
}

TYPED_TEST(ShapeTest, cleaning_path) {
    TypeParam shape;
    auto turns = 100;
    auto seed = 0;
    auto [shuffled, shuffling_path] = shape.shuffle(turns, seed);
    auto unshuffling_path = shuffling_path.reversed();
    auto cleaned_unshuffling_path = shape.clean(unshuffling_path);
    auto reversed_shuffling = shuffling_path;
    std::ranges::reverse(reversed_shuffling);
    for (size_t turn = 0; turn < shuffling_path.size(); turn++) {
        auto [shuffle_move, shuffle_reverse] = shuffling_path.at(turn);
        auto [cleaned_unshuffling_move, cleaned_unshuffling_reverse] = cleaned_unshuffling_path.at(turn);
        auto [reversed_shuffling_move, reversed_shuffling_reverse] = reversed_shuffling.at(turn);
        ASSERT_FALSE(shuffle_reverse);
        ASSERT_FALSE(cleaned_unshuffling_reverse);
        ASSERT_FALSE(reversed_shuffling_reverse);
        ASSERT_EQ(cleaned_unshuffling_move, shape.reverse_of(reversed_shuffling_move));
    }
}
